src/benchmark.rs

```diff
@@ -22,73 +22,228 @@ macro_rules! benchmark_print {
+            counters: [0; NB_COUNTER],
         }
     }
 
-    /// Start benchmarking.
-    pub fn start(&mut self) {
-        if self.is_running {
-            log::error!("already counting instr");
-        } else {
-            self.is_running = true;
+    /// Reset counter value to default and return previous one.
+    fn reset(&mut self, counter: Either) -> usize {
+        let value = counter.reset_value();
+        self.set(counter, value)
+    }
 
-            // instruction
-            if config::BENCHMARK_INSTRUCTION {
-                self.previous_instr_count = Arch::read_csr(Csr::Minstret);
-            }
+    /// Increment counter by one. Mainly useful for occurence counters.
+    fn increment(&mut self, counter: Either) {
+        match counter {
+            Either::Counter(c) => self.counters[c as usize] += 1,
+            Either::IntervalCounter(c) => self.interval_counters[c as usize] += 1,
+        }
+    }
 
-            // time
-            if config::BENCHMARK_TIME {
-                unsafe {
-                    asm!(
-                        "rdtime {value}",
-                        value = out(reg) self.previous_timer
-                    );
-                }
+    /// Set value of a counter and return previous value.
+    fn set(&mut self, counter: Either, value: usize) -> usize {
+        match counter {
+            Either::Counter(c) => {
+                let index = c as usize;
+                let previous = self.counters[index];
+                self.counters[index] = value;
+                previous
+            }
+            Either::IntervalCounter(c) => {
+                let index = c as usize;
+                let previous = self.interval_counters[index];
+                self.interval_counters[index] = value;
+                previous
             }
         }
     }
 
-    /// Stop benchmarking.
-    pub fn stop(&mut self, tag: &str) {
-        self.is_running = false;
+    /// Read value of a counter.
+    fn read(&self, counter: Either) -> usize {
+        match counter {
+            Either::Counter(c) => self.counters[c as usize],
+            Either::IntervalCounter(c) => self.interval_counters[c as usize],
+        }
+    }
 
-        // instruction
-        if config::BENCHMARK_INSTRUCTION {
-            let stop_value = Arch::read_csr(Csr::Minstret);
-            Self::record_entry("instr", tag, stop_value - self.previous_instr_count);
+    /// Reset interval counters.
+    pub fn start_interval_counters() {
+        if !config::BENCHMARK {
+            return;
         }
 
-        // time
-        if config::BENCHMARK_TIME {
-            let stop_time: usize;
-            unsafe {
-                asm!(
-                    "rdtime {value}", 
-                    value = out(reg) stop_time);
+        for counter in [
+            IntervalCounter::ExecutionTime,
+            IntervalCounter::InstructionRet,
+        ]
+        .map(Either::IntervalCounter)
```

Comment:
Wow I didn't know we could do that in Rust, thanks for the trick :)
That implies that `Either::IntervalCounter` is a function, which is cool.